pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -422,11 +434,76 @@ if [ "$action" = dry-build ]; then
     extraBuildFlags+=(--dry-run)
 fi
 
+if [ "$action" = list-generations ]; then
+    if [ ! -L "$profile" ]; then
+        echo "No profile \`$(basename $profile)' found" >&2
+        exit 1
+    fi
+
+    generation_from_dir() {
+        generation_dir="$1"
+        generation_base="$(basename "$generation_dir")" # Has the format "system-123-link" for generation 123
+        no_link_gen="${generation_base%-link}"  # remove the "-link"
+        echo ${no_link_gen##*-} # remove everything before the last dash
+    }
+    describe_generation(){
+        generation_dir="$1"
+        generation_number="$(generation_from_dir "$generation_dir")"
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules")"
```

Comment:
This might not exist in the unlikely case where the kernel is compiled without modules, so it probably should at least have `|| :` so that it doesn't exit.

Reply:
Should it be empty or default to `unknown`? My first instinkt was unknown, but empty might be more appropriate, the more I think about it :thinking:

Reply:
Now you got me thinking: the install-grub skript [seems to assume](https://github.com/NixOS/nixpkgs/blob/6e2f594665a0ddd36b1cb764cb42474a0a8f08e5/nixos/modules/system/boot/loader/grub/install-grub.pl#L543-L546) `lib/modules` exists if `nixos-version` isn't there. Is that a valid assumption?

Reply:
Something is strange in the grub-script: if `nixos-version` doesn't exist it will [skip this generation](https://github.com/NixOS/nixpkgs/blob/6e2f594665a0ddd36b1cb764cb42474a0a8f08e5/nixos/modules/system/boot/loader/grub/install-grub.pl#L538-L540), so the [later check](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L544) should always succed, right? :thinking:

Reply:
Hmm it seems both the grub and the systemd-boot builders assume that the modules directory exists. The code that [assembles the modules](https://github.com/xaverdh/nixpkgs/blob/8aaba2790df85abff58bc7815076aef84743b349/pkgs/build-support/kernel/modules-closure.sh#L8) on the other hand takes extra care to handle this case.
I wonder if there is a better way to pass around the kernel version.

Reply:
and you are right, it should always succeed..

Reply:
[here](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L507) is another instance which isn't guarded, but I have read that perl is a bit relaxed, when it comes to `array out of bound` situations, so I guess it is fine

Reply:
So the actual thing that gets linked in toplevel is `config.system.modulesTree` which is assembled from the `kernel` and `config.boot.extraModulePackages` via the [aggregator](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kmod/aggregator.nix). This also checks if the modules directory exists, so in theory it *can* be missing I think. Probably not too many people compile kernels without modules.

Reply:
Ah never mind, I was confused again. Thats the `kernel-modules` symlink. The bootloader generators read the actual `kernel` symlink..

Reply:
> [here](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L507) is another instance which isn't guarded, but I have read that perl is a bit relaxed, when it comes to `array out of bound` situations, so I guess it is fine?

not sure, I am no expert on perl

Reply:
I think as far as this pr is concerned, you can just fall back to "Unknown" (or something like that) for now.